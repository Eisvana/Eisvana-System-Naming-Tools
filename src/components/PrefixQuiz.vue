<script setup lang="ts">
import type { QuizSystem } from '@/types/types';
import { getRandomGlyphs } from '@/logic/glyphs';
import { glyphs2Coords } from '@/logic/coordGlyphConvert';
import { getPrefix } from '@/logic/prefix';
import { computed, ref } from 'vue';

const quizSystem = ref<QuizSystem>(generateQuizSystem());

function generateQuizSystem(): QuizSystem {
  const glyphs = getRandomGlyphs();
  return {
    glyphs,
    inputName: '',
    coordinates: glyphs2Coords(glyphs),
    expectedPrefix: getPrefix(glyphs),
    clickedSubmit: false,
    solved: false,
  };
}

const newQuizSystem = () => (quizSystem.value = generateQuizSystem());

const isPrefixCorrect = computed(() => quizSystem.value.inputName.startsWith(quizSystem.value.expectedPrefix));

function submitSystemName() {
  if (quizSystem.value.solved) {
    newQuizSystem();
    return;
  }
  quizSystem.value.clickedSubmit = true;
  quizSystem.value.solved =
    isPrefixCorrect.value && quizSystem.value.inputName.length >= quizSystem.value.expectedPrefix.length + 2;

  // delay in milliseconds
  const delay = 3000;
  setTimeout(() => {
    quizSystem.value.clickedSubmit = false;
  }, delay);
}

const output = computed(() => {
  if (!quizSystem.value.solved) {
    return isPrefixCorrect.value ? 'Enter Prefix + Name!' : 'Incorrect!';
  }
  quizSystem.value.solved = true;
  return 'Correct!';
});
</script>

<template>
  <h2>Naming Quiz</h2>
  <p class="cta">Provide an appropriate system name for this address:</p>
  <div class="address">
    <p>
      <span class="has-text-weight-bold">Coordinates:</span>
      <span class="text-size-bigger">{{ quizSystem.coordinates }}</span>
    </p>
    <p>
      <span class="has-text-weight-bold">Glyphs:</span>
      <span class="glyphs">{{ quizSystem.glyphs }}</span>
    </p>
  </div>
  <form @submit.prevent>
    <div class="system-name-input">
      <label for="quizSystemName">Full System Name:</label>
      <input
        v-model="quizSystem.inputName"
        id="quizSystemName"
        placeholder="EVx-xx Hello World"
        type="text"
      />
    </div>
    <div class="actions">
      <button
        type="submit"
        @click="submitSystemName"
      >
        {{ quizSystem.solved ? 'Next' : 'Submit' }}
      </button>
      <button
        v-show="!quizSystem.solved"
        class="secondary"
        type="button"
        @click="newQuizSystem"
      >
        Skip
      </button>
    </div>
  </form>
  <Transition>
    <p
      v-if="quizSystem.clickedSubmit"
      class="status-wrapper"
    >
      <output
        :class="{ success: quizSystem.solved, error: !quizSystem.solved }"
        class="status"
      >
        {{ output }}
      </output>
    </p>
  </Transition>
</template>

<style scoped>
.cta {
  margin-block-end: 1rem;
}

.address {
  user-select: none;
  margin-block-end: 1rem;

  p {
    display: flex;
    align-items: center;
    gap: 0.5rem;
    margin-block-end: 0;
  }

  .glyphs {
    word-break: unset;
    font-size: 2rem;
    line-height: normal;
  }
}
.has-text-weight-bold {
  font-weight: bold;
}

.text-size-bigger {
  font-size: 1.5rem;
}

form {
  .system-name-input {
    width: min-content;
  }

  .actions {
    display: flex;
    flex-wrap: wrap;
    gap: 0.5rem;

    button {
      width: auto;
    }
  }

  input {
    width: auto;
  }
}

.v-leave-active {
  transition: opacity 0.5s ease;
}

.v-leave-to {
  opacity: 0;
}
</style>
