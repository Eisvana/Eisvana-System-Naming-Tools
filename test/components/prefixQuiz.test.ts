import PrefixQuiz from '@/components/PrefixQuiz.vue';
import { vi, beforeEach, describe, expect, it } from 'vitest';
import { mount, type VueWrapper } from '@vue/test-utils';

describe('PrefixQuiz', () => {
  let wrapper: VueWrapper;

  beforeEach(() => {
    wrapper = mount(PrefixQuiz);
    vi.mock('@/logic/glyphs', () => ({
      getRandomGlyphs: vi.fn().mockReturnValue('0001A21117FF'),
    }));
  });

  it('should render correctly', () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should show confirmation if input is correct', async () => {
    const input = wrapper.find('#quizSystemName');
    const confirmBtn = wrapper.find('button[type="submit"]');

    await input.setValue('EV1-1 Hello World');
    await confirmBtn.trigger('click');

    const status = wrapper.find('.status');

    expect(status).toBeDefined();
    expect(status.classes('success')).toBe(true);
    expect(confirmBtn.text()).toBe('Next');
  });

  it('should show error when input is incorrect', async () => {
    const input = wrapper.find('#quizSystemName');
    const confirmBtn = wrapper.find('button[type="submit"]');

    await input.setValue('EV1-5');
    await confirmBtn.trigger('click');

    const status = wrapper.find('.status');

    expect(status).toBeDefined();
    expect(status.classes('error')).toBe(true);
  });
});
