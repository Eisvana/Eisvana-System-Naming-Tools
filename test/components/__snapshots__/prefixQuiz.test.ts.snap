// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`PrefixQuiz > should render correctly 1`] = `
"<h2 data-v-73aa79bf="">Naming Quiz</h2>
<p data-v-73aa79bf="" class="cta">Provide an appropriate system name for this address:</p>
<div data-v-73aa79bf="" class="address">
  <p data-v-73aa79bf=""><span data-v-73aa79bf="" class="has-text-weight-bold">Coordinates:</span><span data-v-73aa79bf="" class="text-size-bigger">0FFE:0021:0910:0001</span></p>
  <p data-v-73aa79bf=""><span data-v-73aa79bf="" class="has-text-weight-bold">Glyphs:</span><span data-v-73aa79bf="" class="glyphs">0001A21117FF</span></p>
</div>
<form data-v-73aa79bf="">
  <div data-v-73aa79bf="" class="system-name-input"><label data-v-73aa79bf="" for="quizSystemName">Full System Name:</label><input data-v-73aa79bf="" id="quizSystemName" placeholder="EVx-xx Hello World" type="text"></div>
  <div data-v-73aa79bf="" class="actions"><button data-v-73aa79bf="" type="submit">Submit</button><button data-v-73aa79bf="" class="secondary" type="button"> Skip </button></div>
</form>
<transition-stub data-v-73aa79bf="" appear="false" persisted="false" css="true">
  <!--v-if-->
</transition-stub>"
`;
